lib/ui/src/modules/ui/libs/filters.js

```diff
@@ -12,10 +12,24 @@ export function storyFilter(stories, filter, selectedKind, sortStoriesByKind) {
   const sorted = sort(stories, sortStoriesByKind);
   if (!filter) return sorted;
 
-  return sorted.filter(kindInfo => {
-    if (kindInfo.kind === selectedKind) return true;
+  return sorted.reduce((acc, kindInfo) => {
+    // Don't filter out currently selected filter
+    if (kindInfo.kind === selectedKind) return acc.concat(kindInfo);
     const needle = filter.toLocaleLowerCase();
     const hstack = kindInfo.kind.toLocaleLowerCase();
-    return fuzzysearch(needle, hstack);
-  });
+
+    // If a match is found in the story hierachy structure return kindInfo
+    if (fuzzysearch(needle, hstack)) return acc.concat(kindInfo);
+
+    // Now search at individual story level and filter results
+    const matchedStories = kindInfo.stories.filter(story => fuzzysearch(needle, story));
```

Comment:
IMO you should add the lowercasing here (like for the `kind` check, looks like `fuzzysearch` already do `===`, so it maybe redundant) And add tests for this usecase.